src/x11.rs

```diff
@@ -131,33 +137,170 @@ impl X11Impl {
             .reply()
             .swbuf_err("Failed to get geometry reply")?;
 
+        // See if SHM is available.
+        let shm_info = {
+            let present = query_token
+                .reply()
+                .swbuf_err("Failed to get SHM query reply")?
+                .present;
+
+            if present {
+                // SHM is available.
```

Comment:
Well... no. This only checks whether the SHM *extension* is available. But this does not check whether the X11 server and this process run on the same machine.... I think.

Are you sure the X11 server hides the SHM extension for non-local clients? How would that even work with ssh X11 forwarding? The X11 server would not know that ssh is forwarding something non-local, would it?

Anyway, cairo does a lot more complicated things to figure out whether SHM is usable: https://gitlab.freedesktop.org/cairo/cairo/-/blob/13a0474641232cccb15d24fa31d8c2557731665b/src/cairo-xcb-connection.c#L443-479

This seems to set up some shared memory and actually check whether a SHM Attach and Detach request succeed.

Reply:
I used something similar that sets up shared memory and then attaches/detaches a buffer. This is probably something that we should split into the display if/when we do #37